src/main/java/com/google/sps/servlets/NewsServlet.java

```diff
@@ -0,0 +1,114 @@
+
+  public final String GOOGLE_NEWS_PATH = "https://news.google.com/search";
+  public final String articleTag = "DY5T1d";  // Class tag for articles in Google News HTML code.
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    List<String> terms = Arrays.asList(request.getParameterValues("key"));
+    HashMap<String, List<Article>> articleMap = new HashMap<>();
+    terms.forEach((term) -> {
+      try {
+        Map<String, String> paramMap = new HashMap<>();
+        paramMap.put("q", UrlRequest.encodeTerm(term));
+        paramMap.put("hl", "en-US");
+        paramMap.put("gl", "US");
+        paramMap.put("ceid", "US:en");
+        HttpURLConnection connect = getConnection(GOOGLE_NEWS_PATH, paramMap);
+        String HTMLString = getHTML(connect);
+        List<Article> articleList = makeArticleList(HTMLString);
+        articleMap.put(term, articleList);
+      }
+      catch (IOException e) {
+        e.printStackTrace();
+      }
+    });
+    String json = encodeMapAsJson(articleMap);
+    response.setCharacterEncoding("UTF-8");
+    response.getWriter().println(json);
+  }
+
+  public HttpURLConnection getConnection(String basePath, Map<String, String> paramMap) throws IOException {
+    String path = String.format("%s?%s", basePath, UrlRequest.getParamsString(paramMap));
+    URL url = new URL(path);
+    HttpURLConnection connection = (HttpURLConnection) url.openConnection();
+    connection.setRequestMethod("GET");
+    connection.setRequestProperty("Accept", "text/html");
+    connection.setRequestProperty("Content-Type", "text/html");
+    int responseCode = connection.getResponseCode();
+    if (responseCode != 200) {
+      System.err.println("Error: connection response code is: " + responseCode);
+    }
+    return connection;
+  }
+
+  public String getHTML(HttpURLConnection connection) throws IOException {
+    InputStreamReader inputStream = new InputStreamReader(connection.getInputStream());
+    StringBuilder contentBuilder = new StringBuilder();
+    try (BufferedReader inputReader = new BufferedReader(inputStream)) {
+      String inputLine;
+      while ((inputLine = inputReader.readLine()) != null) {
+        contentBuilder.append(inputLine);
+      }
+    } catch(IOException e) {
+      System.err.println(e);
+      return "";
+    } finally {
+      connection.disconnect();
+    }
+    return contentBuilder.toString();
+  }
+
+  public List<Article> makeArticleList(String HTMLString) throws IOException {
+    List<Article> articleList = new ArrayList<>();
+    String[] substrings = HTMLString.split(articleTag);
+    String thisSubstring, lastSubstring, title, link;
+    int closeBracket, openBracket;
+    // For each substring seperated by the article class tag.
+    for (int i = 5; i < substrings.length; i++) {
+      // Get the link from before the article tag and the title from after to make a new Article object.
```

Comment:
[Optional] Giving an example of what the substrings look like might help understand made the code clearer.